compiler/src/iree/compiler/Codegen/Common/ResolveWorkgroupCountHints.cpp

```diff
@@ -0,0 +1,1043 @@
+      rewriter, map, workloadVals, sliceMap[callSlice.callee], sliceMap);
+  if (failed(sizes)) {
+    return failure();
+  }
+
+  // Erase the arg mapping for the function args. Each call to a function
+  // provides a different mapping and may have a different set of optional
+  // args.
+  for (Value v : argsToUnmap) {
+    map.erase(v);
+  }
+
+  assert(sizes->size() == 3 &&
+         "function materialization produced too few values");
+
+  if (placeholder) {
+    Operation *parent = placeholder->getParentOp();
+    SmallVector<Value, 3> newSizes = parent->getResults();
+    rewriter.replaceOp(placeholder, *sizes);
+    *sizes = newSizes;
+    rewriter.setInsertionPointAfter(parent);
+  }
+
+  return *sizes;
+}
+
+/// Materializes the IR for all contained hint and required call ops at the
+/// current insertion point and then takes the maximum values of each
+/// elementwise. For example:
+///
+/// ```
+/// func.func @entry_point() {
+///   iree_codegen.workgroup_count_hint %x, %y, %z
+///   iree_codegen.workgroup_count_hint %a, %b, %c
+///   func.call @constant_hint()
+/// }
+/// func.func @constant_hint() {
+///   iree_codegen.workgroup_count_hint 1, 2, 3
+/// }
+/// ```
+///
+/// This would resolve to:
+///
+/// ```
+/// hal.executable.export @entry_point count(%x, %y, %z, %a, %b, %c) {
+///   %wx = max(%x, %a, 1)
+///   %wy = max(%y, %b, 2)
+///   %wz = max(%z, %c, 3)
+///   hal.return %wx, %wy, %wz
+/// }
+/// ```
+static FailureOr<SmallVector<Value, 3>>
+materializeFunc(RewriterBase &rewriter, IRMapping &map, ValueRange workloadVals,
+                FunctionSlice &funcSlice,
+                llvm::DenseMap<FunctionOpInterface, FunctionSlice> sliceMap) {
+  SmallVector<Value, 3> results;
+  for (HintSlice &hint : funcSlice.hints) {
+    FailureOr<SmallVector<Value, 3>> newVals =
+        materializeHint(rewriter, map, workloadVals, hint);
+    if (failed(newVals)) {
+      return failure();
+    }
+
+    if (results.empty()) {
+      results = newVals.value();
+    } else {
+      results = maxValueVectors(rewriter, hint.hintOp.getLoc(), results,
+                                newVals.value());
+    }
+  }
+
+  for (CallSlice &call : funcSlice.calls) {
+    FailureOr<SmallVector<Value, 3>> newVals =
+        materializeCall(rewriter, map, workloadVals, call, sliceMap);
+    if (failed(newVals)) {
+      return failure();
+    }
+
+    if (results.empty()) {
+      results = newVals.value();
+    } else {
+      results = maxValueVectors(rewriter, call.callee.getLoc(), results,
+                                newVals.value());
+    }
+  }
+
+  if (results.empty()) {
+    // Something went wrong with error catching during analysis or
+    // materialization.
+    return failure();
+  }
+  return results;
+}
+
+/// Replaces the given `workgroup_count_from_slice` op with the program slice
+/// described by the FunctionSlice |root|.
+static LogicalResult replaceFromSliceOpWithFunctionSlice(
```

Comment:
Seems strange to me that this is taking the entire `sliceMap` as an argument. Also you probably want to pass that as reference.